### PR TITLE
Response Time Breakdowns by HTTP Status Code

### PR DESCRIPTION
**Pull request by jeremyandrews**
## Summary
This PR implements response time breakdowns grouped by HTTP status code in Goose's CLI and HTML/PDF reports, providing valuable performance insights for load testing analysis.

Closes https://github.com/tag1consulting/goose/issues/528

## What's New
When a request returns multiple different HTTP status codes during a load test, Goose now automatically provides response time breakdowns grouped by status code. This feature uses smart omission - breakdowns only appear when multiple status codes exist, preventing unnecessary clutter.

### CLI Output Example
```
GET static asset         |        5.11 |          2 |          38 |          5
  └─ 200 (94.2%)         |        4.98 |          2 |          35 |          5
  └─ 404 (5.8%)          |        7.24 |          5 |          38 |          7
```

This breakdown reveals that 404 responses took longer on average (7.24ms) than successful 200 responses (4.98ms), helping identify performance patterns related to error handling.

### HTML Report Integration
The feature is fully integrated into HTML and PDF reports with professional formatting:
- Clean visual hierarchy with proper indentation
- Tree-like formatting clearly indicates sub-metrics
- Works seamlessly with existing report generation

## Technical Implementation
- **Zero configuration**: Works automatically when status code tracking is enabled
- **Smart omission**: Only shows breakdowns for requests with multiple status codes
- **Memory efficient**: Leverages existing data structures
- **Backward compatible**: No breaking changes to existing functionality
- **Comprehensive test coverage**: New integration tests ensure reliability

## Files Modified
- `src/metrics.rs`: Core tracking infrastructure
- `src/metrics/common.rs`: HTML report generation
- `src/report.rs`: Professional HTML formatting
- `tests/status_code_response_times.rs`: Comprehensive test coverage
- `src/docs/goose-book/src/getting-started/metrics.md`: Updated documentation

## Benefits
- **Performance insights**: Identify response time differences between success/error responses
- **Error analysis**: Spot performance patterns in error handling
- **Optimization targeting**: Focus efforts on specific response types
- **Professional reporting**: Clean HTML/PDF output suitable for stakeholders

### CODE REVIEW

File: src/metrics.rs [hunk 1]
```diff
@@ -1845,6 +1868,50 @@ impl GooseMetrics {
...
+                    if let Some(timing_data) = request.status_code_response_times.get(&status_code)
+                    {
+                        // Calculate percentage of requests with this status code
+                        let status_count =
+                            request.status_code_counts.get(&status_code).unwrap_or(&0);
+                        let total_count = request.success_count + request.fail_count;
+                        let percentage = if total_count > 0 {
+                            (*status_count as f32 / total_count as f32) * 100.0
+                        } else {
+                            0.0
```
[the remark this answers is not in the document]
Comment (jeremyandrews) [reply src/metrics.rs:1888]: Used .copied().unwrap_or(0) instead of .unwrap_or(&0) with dereference for cleaner, more idiomatic Rust code when accessing HashMap values.

**Review by LionsAd — COMMENTED** [review]
## PR Review: Response Time Breakdowns by HTTP Status Code (#663)

### **🎯 Core Assessment**
This is a **valuable feature** with a **solid implementation** that provides real diagnostic benefits. The concerns raised are legitimate but represent different levels of priority and business tradeoffs.

### **📋 Recommendations by Category**

#### **🔧 Technical Improvements (If Time Permits)**

**Memory Optimization**
- **Problem**: Storing duplicate timing data per status code
- **Options**: 
  - Add config flag to disable feature for high-RPS scenarios
  - Use aggregate stats only (min/max/count) instead of full vectors
  - Document the memory tradeoff in user docs
- **Business Reality**: Most load tests have reasonable request volumes where this won't matter

**API Design Consideration**
- **Problem**: Adding public field to public struct 
- **Options**:
  - Make field `pub(crate)` with accessor methods
  - Accept as minor version bump (common in ecosystem)
  - Add `#[non_exhaustive]` for future flexibility
- **Business Reality**: Many Rust crates evolve public structs this way

#### **🧪 Test Reliability (Quick Wins)**

**Mock Determinism**
- **Problem**: httpmock behavior isn't guaranteed to hit all mocks
- **Solution**: Use stateful mocking or explicit ratios
- **Effort**: Low - just restructure existing test setup

**Edge Case Safety**
- **Problem**: Empty vectors in median calculation
- **Solution**: Add guard clauses or verify `util::median` handles empty input
- **Effort**: Very low - add a few safety checks

#### **🎨 Polish (Nice to Have)**

**HTML Accessibility**
- **Problem**: Hardcoded background color
- **Solution**: Use CSS variables or classes
- **Business Reality**: Current approach works fine, this is enhancement

**Code Style**
- **Problem**: Clippy warnings
- **Solution**: Run `cargo clippy --fix`
- **Effort**: Literally automatic

(Powered by Claude Code)

**Review by LionsAd — APPROVED** [review]
RTBM - with some concerns raised by the code review

I leave it to your discretion if you want to implement them or not or want to create follow-up issues.

**Comment by jeremyandrews** [comment]
> Here is the clippy diff - it must be my version, which is still overly pedantic, but I also find it easier to read:

LIkely you need to update your local version of Rust and related tools. https://github.com/rust-lang/rust-clippy/blob/master/CHANGELOG.md  The `uninlined_format_args` lint was enabled by default in our configuration for a while, but then re-disabled. While we could force re-enable it, I'd prefer to wait for it to be a clippy default.
